vignettes/use-application-audit.Rmd

```diff
@@ -0,0 +1,492 @@
+In practice, we'll have to:
+
+- Start the Shiny app as a background process on a given port. 
+- Fire the load test recorder on port `8600`.
+- Connect Chrome to the recorder on port `8600`. 
+
+### Launch the background app
+
+You may have already noticed that when launching a Shiny app, you can’t run 
+anything else in the R console while the app is live. The explanation is pretty 
+simple: since R is single threaded, it can only use one processor at a time. 
+In practice, this means that tasks are run sequentially. 
+
+How do we start the app without blocking the main R process? 
+We leverage the `{callr}` package, which exposes a convenient API to start R 
+processes in the background, that is, without blocking the main R process. 
+Below shows the code to start a Shiny app located at `path` on a specific port 
+and run the load test recorder on the same port:
+
+``` r
+# Main shiny app
+shiny_bg <- function(path, port) {
+ # main app
+  options(shiny.port = port)
+  shiny::runApp(path)
+}
+
+# Start recorder
+recorder_bg <- function(port) {
+  shinyloadtest::record_session(
+    target_app_url = sprintf("http://127.0.0.1:%s", port),
+    host = "127.0.0.1",
+    port = 8600,
+    output_file = "recording.log",
+    open_browser = FALSE
+  )
+}
+```
+
+We can pass this to the `start_r_bg()` function:
+
+```r
+start_r_bg <- function(fun, path = NULL, port = 3515) {
+ 
+  # remove NULL elements
+  args <- Filter(Negate(is.null), list(path = path, port = port))
+
+  process <- callr::r_bg(
+    func = fun,
+    args = args,
+    stderr= "",
+    stdout = ""
+  )
+
+  while (any(is.na(pingr::ping_port("127.0.0.1", 3515)))) {
+    message("Waiting for Shiny app to start...")
+    Sys.sleep(0.1)
+  }
+
+  attempt::stop_if_not(
+    process$is_alive(),
+    msg = "Unable to launch the subprocess"
+  )
+
+  process
+}
+```
+
+where `r_bg()` starts a background R process, passing the corresponding function 
+and parameters. Besides, we provide some log elements and safety guard in case 
+the app can't start. To launch the app and recorder we simply call:
+
+``` r
+target <- start_r_bg(shiny_bg, path = "vignettes/non-optimized-app/app.R")
+# Listening on 127.0.0.1:3515
+recorder <- start_r_bg(recorder_bg)
+# Listening on 127.0.0.1:8600
+```
+
+### Connect Chrome
+
+The previous part was the most technical step. Now, we only have to start a 
+Chrome headless browser on port `8600`, where the load test recorder runs. 
+You'll notice that `{shinytest2}` also supports remote urls. 
+We ensure to set a consistent load timeout, which will avoid to see this kind of 
+annoying `{shinytest2}` warning:
+
+```
+{shinytest2} R  info 15:25:37.36 Error while initializing AppDriver:
+Shiny app did not become stable in 10000ms.
+```
+
+``` r
+# start headless chrome (points to recorder!).
+# AppDriver also support remote urls.
+chrome <- shinytest2::AppDriver$new("http://127.0.0.1:8600", load_timeout = 15000)
```

Comment:
Spread arg calculation so it can easily be read ; Also changing name to `app`

```suggestion
chrome <- shinytest2::AppDriver$new("http://127.0.0.1:8600", load_timeout = 15 * 1000)
```